### PR TITLE
WriteDemo - 'old' front-end only

### PR DESCRIPTION
**Pull request by jonmattgray**
Front-end only username and password input in the 'old' style. An array of `[username, password]` is stored in context. Appends the new pair each time a user tries to login. No sending to API yet - for next PR.

Includes:
- tests
- update all tests to use `data-cy` selector

image
image
image

### CODE REVIEW

File: src/components/demos/write/Demo.js [hunk 1]
```diff
@@ -1,7 +1,143 @@
...
+        writeDemo: {
+          ...writeDemo,
+          usernamePasswordPairs: [
+            ...writeDemo.usernamePasswordPairs,
+            [usernameInput, passwordInput],
+          ],
+        },
+      })
+      // TODO execute API
+      SetDemoOutput('***PARSED API OUTPUT***')
```
Comment (mattdean-digicatapult) [inline src/components/demos/write/Demo.js:84]: I think this needs some refinement. Perhaps on an unsuccessful login attempt we could display an error in this area and then on a successful login make the login screen disappear and present a desktop?

File: src/components/demos/write/Demo.js [hunk 2]
```diff
@@ -1,7 +1,144 @@
...
+  const [someUsernameTyped, SetSomeUsernameTyped] = useState(false)
+  const [somePasswordTyped, SetSomePasswordTyped] = useState(false)
+
+  const noUsernameEntered = usernameInput.length === 0 && someUsernameTyped
+  const usernameAtMaxLength = usernameInput.length === usernameUpperBound
+  const noPasswordEntered = passwordInput.length === 0 && somePasswordTyped
+  const passwordAtMaxLength = passwordInput.length === passwordUpperBound
+
+  const UsernameErrorWarning = () => (
+    <>
```
Comment (n3op2) [inline src/components/demos/write/Demo.js:42]: do we need <React.Fragment>? Could it be just an array?

File: src/components/demos/write/Demo.js [hunk 3]
```diff
@@ -1,7 +1,144 @@
...
+    }
+  }, [usernameInput, passwordInput])
+
+  useEffect(() => {
+    update(initState)
+  }, [update])
+
+  const enterUsernameAndPassword = (e) => {
+    e.preventDefault()
+    SetSomeUsernameTyped(true)
```
Comment (n3op2) [inline src/components/demos/write/Demo.js:70]: Not sure if I like setter so be in capital, it's not a component or class.

**Review by n3op2 — APPROVED** [review]

File: src/components/demos/write/Demo.js (hunk 3 above)
Comment (jonmattgray) [reply src/components/demos/write/Demo.js:70]: sorry for some reason I got it into my head that setters start with capital. Will change

File: src/components/demos/write/Demo.js (hunk 2 above)
Comment (jonmattgray) [reply src/components/demos/write/Demo.js:42]: I figured both warnings can never happen at the same time so may as well use a fragment rather than list

**Review by mattdean-digicatapult — APPROVED** [review]

**Pull request merged**